### ISSUE TITLE
How to Set/Change layer/model name

### ISSUE THREAD

**Issue by Honzys** [keras-team/keras#6194]
Hello,
how can I set the name of Sequential model, or how can I change the layer name?
Let's say I have this model:
```
def create_base_network(input_dim):
    model = Sequential(name="base")
    model.add(Convolution2D(32, 3, 3, border_mode='valid', name='conv1_1',
                            input_shape=input_dim))
    model.add(BatchNormalization(mode=2, axis=-1))
    model.add(Activation('relu'))
    model.add(Convolution2D(32, 3, 3, border_mode='valid', name='conv1_2'))
    model.add(BatchNormalization(mode=2, axis=-1))
    model.add(Activation('relu'))
    return model
```
Then I create siamese network, with upper model as core:
```
input_a = Input(shape=input_dim)
input_b = Input(shape=input_dim)

base_network = create_base_network(input_dim)

processed_a = base_network(input_a)
processed_b = base_network(input_b)

distance = Lambda(euclidean_distance, output_shape=eucl_dist_output_shape)([processed_a, processed_b])

model = Model(input=[input_a, input_b], output=distance)
```
Then I save the model, and If I try to load the model and get the layer by name `"base"`, there's no layer. There's only `"sequential_1"`.

So my question is, where should I put the `name="base"` attribute if I want to get the "core" layer?

Or is there any way to rename the layer? I tried `model.layers[2].name = "my_desired_name"`.
But If I save the model and then load again, there's error with loading,

I'd be glad for any help or advice how to solve this.
Thanks anyway!

**Comment by dc-ai** [comment]
I would like to know about this too. I realized you can't load two models and form a neural network, as Keras errors out because the two models have the same name.

**Comment by Narts** [comment]
Hi, did you figure it out?

**Comment by divinekage** [comment]
how to name model ?

**Comment by xinghuokang** [comment]
It is easy! Just rename the model when you load model, like this 
model_sf = load_my_model('conv-net2.h5', 'conv-net2.json')
model_sf.name = 'model_1'  # "model_1 : rename as you like"

**Comment by shahabaslani** [comment]
@Narts Did you find any solution?!

**Comment by KiraMichiru** [comment]
I would also be interested in a solution, as this still seems to be an issue. While the model summary shows me model names (used as layers in a bigger model) correctly before saving the model, the names of the layers that have a Sequential model underlying are not preserved when saving and re-loading. The names of layers having other models (not Sequential) underlying, however, are preserved.

**Comment by shpavisaslan** [comment]
I used the following code and it is working for me.

# Model
I1 = Input(shape=(304, 304, 3))
model1 = ResNet50(weights='imagenet', include_top= False, input_tensor= I1, pooling=None)
model1.layers.pop()
model1.outputs = [model1.layers[-1].output]
model1.layers[-1].outbound_nodes = []
output1 = Model(inputs=I1, outputs=[model1.get_layer('activation_49').output])

I2 = Input(shape=(304, 304, 3))
model2 = ResNet50(weights='imagenet', include_top= False, input_tensor= I2, pooling=None)
for layer in model2.layers:
    layer.name = layer.name + str("_2")
model2.layers.pop()
model2.outputs = [model2.layers[-1].output]
model2.layers[-1].outbound_nodes = []
output2 = Model(inputs=I2, outputs=[model2.get_layer('activation_98_2').output])
...

**Comment by adizhol** [comment]
@shpavisaslan 
I'm getting _AttributeError: can't set attribute_ when trying to set layer.name
Any idea why?

tensorflow 1.4.0
keras 2.0.8-tf
win10

**Comment by Schlampig** [comment]
The following is a stupid way I used to implement **Siamese CNN**, but it works:
# The trick is to give different name to the output layers and never name any other layers

def get_body(x_input=None, img_shape=(256, 128, 3), name=''):
    assert x_input is not None
    x = Conv2D(64, (10, 10),
               activation='relu', input_shape=img_shape,
               kernel_initializer=W_init,
               kernel_regularizer=l2(2e-4))(x_input)
    x = MaxPooling2D()(x)
    ......
    x_fea = Flatten()(x)
    x_output = Dense(4096,
                     activation="sigmoid",
                     **name=name+'output_layer'**)(x_fea)
    return x_output

left_head = Input(shape=img_shape)
left_body = get_body(x_input=left_head, img_shape=img_shape, **name='left_'**)
right_head = Input(shape=img_shape)
right_body = get_body(x_input=right_head, img_shape=img_shape, **name='right_'**)
body = Lambda(lambda x: K.abs(x[0] - x[1]))([left_body, right_body])
spirit = Dense(1, activation='sigmoid', bias_initializer=b_init)(body)
life = Model(inputs=[left_head, right_head], outputs=spirit)

**Comment by iretiayo** [comment]
You can add an identity layer and name it accordingly  e.g:
```
model.add(Lambda(lambda x: x, name="new_layer_name"))
```
The identity layer simply returns as output whatever went in.

**Comment by baregawi** [comment]
Since your Lambda has no trainable parameters, you can name it any random thing and it will behave the same every time.
@Schlampig I didn't have to do this for every layer in the neural net, just the Lambdas. It would seem limiting or even harmful to not share parameters for two sides of a siamese network. I just added a random name to each Lambda layer since my Lambdas were just mathematical functions with no trainable parameters and would work the same regardless of identity.

**Comment by AnkBurov** [comment]
@adizhol just found out how to resolve `can't set attribute` error - use 
`model._name = '!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!'`
Just checked, new model name is persisted after model serialization and deserialization, so it's quite a way to go.

**Comment by SoccoCMOS** [comment]
Sequential(layers=[Add your layers here],name="set_name_here")

**Comment by carusyte** [comment]
> @adizhol just found out how to resolve `can't set attribute` error - use
> `model._name = '!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!'`
> Just checked, new model name is persisted after model serialization and deserialization, so it's quite a way to go.

But isn't that the attribute names with a single leading underscore supposed to be internal to the class?

Albeit `model._name` is accessible from outside the class, maybe we need an idiomatic way to set the model name in any of functional API / subclassing / Sequential model.

**Comment by athewsey** [comment]
+1 for a way to rename a layer (and model too, I guess): For a complex model constructed via functional API, the actual creation of the output layer might be buried somewhere in a reusable code block... I wanna override the name after the initial creation; not pass a name parameter down some potentially long chain of building functions.

As @carusyte mentioned, surely `._name` is not intended as an external API: Maybe the `.name` attribute is readonly purely to prevent accidents?

**Comment by batzner** [comment]
How about `model = Model(model.inputs, model.outputs, name='new name')`?

**Comment by nrasadi** [comment]
I just posted a short guide in Medium. Hope you find it helpful.

https://nrasadi.medium.com/change-model-layer-name-in-tensorflow-keras-58771dd6bf1b
